### PR TITLE
ldap3 extended logging when debug mode is enabled

### PR DESCRIPTION
**Pull request by glicht**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19328

## Description
Extended logging when debug-mode=true for ldap3. Also added to common server `is_debug_mode` method.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
CommonServer added `is_debug_mode wrapper method

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [Changelog Active Direcotry](https://github.com/demisto/content/blob/ldap3-debug-mode/Integrations/Active_Directory_Query/CHANGELOG.md)
- [ ] [CHANGELOG common server](https://github.com/demisto/content/blob/ldap3-debug-mode/Scripts/CommonServerPython/CHANGELOG.md)

### CODE REVIEW

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 1]
```diff
@@ -1025,14 +1031,19 @@ def main():
             search_group_members(DEFAULT_BASE_DN, DEFAULT_PAGE_SIZE)
 
     except Exception as e:
-        message = "{}\nLast connection result: {}\nLast error from LDAP server: {}".format(
-            str(e), json.dumps(conn.result), conn.last_error)
+        message = str(e)
+        if conn:
+            message += "\nLast connection result: {}\nLast error from LDAP server: {}".format(
+                json.dumps(conn.result), conn.last_error)
         demisto.info(message)
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: can remove this, `return_error` already runs `demisto.info`

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 2]
```diff
@@ -914,59 +915,64 @@ def main():
...
+        try:
+            set_library_log_hide_sensitive_data(True)
+            if is_debug_mode():
+                demisto.info('debug-mode: setting library log detail to EXTENDED')
+                last_log_detail_level = get_library_log_detail_level()
+                set_library_log_detail_level(EXTENDED)
+            server = initialize_server(SERVER_IP, PORT, SECURE_CONNECTION, UNSECURE)
+        except Exception as e:
+            return_error(str(e))
+            return
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: is this here for the linter?

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 3]
```diff
@@ -914,59 +915,64 @@ def main():
...
-            "Last connection result: {}\n" \
-            "Last error from LDAP server: {}".format(json.dumps(conn.result), json.dumps(conn.last_error))
-        demisto.info(message)
-        return_error(message)
-        return
+        if not base_dn_verified(DEFAULT_BASE_DN):
+            message = "Failed to verify the base DN configured for the instance.\n" \
+                "Last connection result: {}\n" \
+                "Last error from LDAP server: {}".format(json.dumps(conn.result), json.dumps(conn.last_error))
+            demisto.info(message)
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: can remove this, `return_error` already runs `demisto.info`

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 4]
```diff
@@ -914,59 +915,64 @@ def main():
...
+        else:
+            # here username should be the user dn
+            conn = Connection(server, user=USERNAME, password=PASSWORD)
+
+        # bind operation is the “authenticate” operation.
+        try:
+            # open socket and bind to server
+            if not conn.bind():
+                message = "Failed to bind to server. Please validate the credentials configured correctly.\n{}".format(
+                    json.dumps(conn.result))
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: can remove this, `return_error` already runs `demisto.info`

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 5]
```diff
@@ -914,59 +915,64 @@ def main():
...
+        except Exception as e:
+            exc_msg = str(e)
+            demisto.info("Failed bind to: {}:{}. {}: {}".format(SERVER_IP, PORT, type(e), exc_msg
+                         + "\nTrace:\n{}".format(traceback.format_exc())))
+            message = "Failed to access LDAP server. Please validate the server host and port are configured correctly"
+            if 'ssl wrapping error' in exc_msg:
+                message = "Failed to access LDAP server. SSL error."
+                if not UNSECURE:
+                    message += ' Try using: "Trust any certificate" option.'
             demisto.info(message)
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: can remove this, `return_error` already runs `demisto.info`

File: Integrations/Active_Directory_Query/Active_Directory_Query.py [hunk 6]
```diff
@@ -914,59 +915,64 @@ def main():
...
+            set_library_log_hide_sensitive_data(True)
+            if is_debug_mode():
+                demisto.info('debug-mode: setting library log detail to EXTENDED')
+                last_log_detail_level = get_library_log_detail_level()
+                set_library_log_detail_level(EXTENDED)
+            server = initialize_server(SERVER_IP, PORT, SECURE_CONNECTION, UNSECURE)
+        except Exception as e:
+            return_error(str(e))
+            return
+        global conn
```
Comment (Itay4) [inline Integrations/Active_Directory_Query/Active_Directory_Query.py]: did we consider setting `fast_decoder=False` in the connection object ?

Comment (glicht) [reply Integrations/Active_Directory_Query/Active_Directory_Query.py]: Don't think so. Why disable it? It seems to be working fine.

Comment (Itay4) [reply Integrations/Active_Directory_Query/Active_Directory_Query.py]: from ldap3 docs:
```
To get a full descriptive logging of outgoing messages you must set fast_decoder=False in the connection object.
```

File: Integrations/Active_Directory_Query/Active_Directory_Query.py (hunk 2 above)
Comment (glicht) [reply Integrations/Active_Directory_Query/Active_Directory_Query.py]: yes. linter doesn't know that this call sys.exit.

**Comment by glicht** [comment]
I'll check regarding a test

**Comment by kirbles19** [comment]
@glicht Done.

**Comment by glicht** [comment]
Test playbook added. Build failure seems ok as a result of running all tests because of CommonServer modification.

image

**Pull request merged**